python/data_structures/linked_list.py

```diff
@@ -1,62 +1,107 @@
 
+"""
+Can successfully instantiate an empty linked list
+Can properly insert into the linked list
+The head property will properly point to the first node in the linked list
+Can properly insert multiple nodes into the linked list
+Will return true when finding a value within the linked list that exists
+Will return false when searching for a value in the linked list that does not exist
+Can properly return a collection of all the values that exist in the linked list
+"""
+
+
 class LinkedList:
-    """
-    Can successfully instantiate an empty linked list
-    Can properly insert into the linked list
-    The head property will properly point to the first node in the linked list
-    Can properly insert multiple nodes into the linked list
-    Will return true when finding a value within the linked list that exists
-    Will return false when searching for a value in the linked list that does not exist
-    Can properly return a collection of all the values that exist in the linked list
-    """
 
     def __init__(self, value='None'):
         self.head = None
         self.value = value
         self.next = next
 
-    def insert(self, value):
+    def includes(self, value):
+        node = self.head
 
+        while node:
+            if node.value == value:
+                return True
+            node = node.next
+        return False
+
+    def insert(self, value):
         new_node = Node(value)
-        new_node.next = self.head
+        node = self.head
+
+        new_node.next = node
         self.head = new_node
 
-    def includes(self, value):
+    def insert_before(self, target_value, value):
+        new_node = Node(value)
+        node = self.head
 
-        current = self.head
-        while current:
-            if current.value == value:
-                return True
-            current = current.next
-        return False
+        if node is None:
+            raise TargetError('There are no Nodes to insert before')
+        if not self.includes(target_value):
+            raise TargetError(f'{{ {target_value} }} does not exist')
+        else:
+            while node.next:
+                if node.next.value == target_value:
+                    new_node.next = node.next
+                    node.next = new_node
+                    break
+            if node.value == target_value:
+                new_node.next = node
+                self.head = new_node
 
-    def __str__(self):
+    def insert_after(self, target_value, value):
+        new_node = Node(value)
+        node = self.head
+
+        while node:
+            if node.value == target_value:
+                new_node.next = node.next
+                node.next = new_node
+                break
+            raise TargetError(f'{target_value} is missing')
+        if node is None:
+            raise TargetError('There are no nodes')
```

Comment:
Your insert_after method does not actually traverse the linked list, and raises an exception if the target value is not the head value.

Revise the insert_after method to traverse the list, and function properly if inserting after a value other than head, such as the second or third value of the linked list.

I recommend creating a new test or two to ensure proper behavior in this case. Seems like a gap in the provided testing suite has been found.